### PR TITLE
 [WIP] Stream base3.0

### PR DESCRIPTION
**Pull request by panxiaojun233**
## What is the purpose of the change

XXXXX

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Message.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package org.apache.dubbo.rpc.protocol.tri;
+
+import java.io.InputStream;
+
+public class Message {
+    private Object headers;
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Message.java:6]: 1. 定义成 Object 不合适，需要有具体的方法
2. 好像没有地方调用getHeaders

File: dubbo-common/src/main/java/org/apache/dubbo/common/stream/StreamObserver.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package org.apache.dubbo.common.stream;
+
+public interface StreamObserver<T> {
+    void onNext(T var1) throws Exception;
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/stream/StreamObserver.java:4]: onNext 不应该抛Exception

File: dubbo-common/src/main/java/org/apache/dubbo/common/stream/StreamObserver.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+    void onError(Throwable var1);
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/stream/StreamObserver.java:6]: 变量名要有具体含义，不要用var/foo/bar  这种

File: dubbo-common/src/main/java/org/apache/dubbo/rpc/model/MethodDescriptor.java [hunk 4]
```diff
@@ -40,18 +46,54 @@
     private final Type[] returnTypes;
     private final String methodName;
     private final boolean generic;
+    private final boolean stream;
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/rpc/model/MethodDescriptor.java:49]: 可以抽象出一个方法类型枚举，将stream 和 needwrap 合成一个字段。stream 都是不需要 wrap 的，unary_wrap/ unary_unwrap 这种

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java [hunk 5]
```diff
@@ -43,11 +44,11 @@
             .withDescription("Too many data");
     private final ChannelHandlerContext ctx;
     private final URL url;
+    private boolean needWrap;
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java:47]: 这里的needwrap 能否改为MethodDescriptor

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java [hunk 6]
```diff
@@ -43,11 +44,11 @@
...
     private MultipleSerialization multipleSerialization;
     private Http2Headers headers;
     private Http2Headers te;
-    private boolean needWrap;
-    private InputStream data;
+    private Queue<InputStream> datas = new ArrayDeque<>();
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java:51]: arrayDeque 会自动扩容，如果出现请求堆积，容易内存泄漏，这里设置一个合理的默认值，再提供参数配置是不是更合理，参考netty的各种参数格式

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java [hunk 7]
```diff
@@ -100,22 +97,26 @@ public Http2Headers getTe() {
         return te;
     }
 
+    public InputStream pollData() {
+        return datas.poll();
+    }
+
     public InputStream getData() {
-        return data;
+        return datas.peek();
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java:105]: 这个方法没用到

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/ClientStream.java [hunk 8]
```diff
@@ -224,4 +224,7 @@ public void halfClose() {
         }
     }
 
+    @Override
+    protected void onSingleMessage(InputStream in) throws Exception {
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/ClientStream.java:228]: TBD?

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Message.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    private InputStream is;
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Message.java:7]: final

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Processor.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import io.netty.channel.ChannelHandlerContext;
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.common.serialize.MultipleSerialization;
+import org.apache.dubbo.rpc.model.MethodDescriptor;
+import org.apache.dubbo.triple.TripleWrapper;
+
+// metadata + is -> object
+// object -> is
+public class Processor {
+    private ServerStream serverStream;
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Processor.java:21]: final

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Processor.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        MultipleSerialization multipleSerialization) {
+        this.serverStream = serverStream;
+        this.md = md;
+        this.url = url;
+        this.serializeType = serializeType;
+        this.multipleSerialization = multipleSerialization;
+    }
+
+    public void onSingleMessage(InputStream in) throws Exception {
+        if (serverStream instanceof StreamServerStream) {
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Processor.java:37]: 用抽象，不要用instance of

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Processor.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            if (resp.length > 1) {
+                return;
+            }
+            stream.getObserver().onNext(resp[0]);
+        }
+
+    }
+
+    public Object[] decodeRequestMessage(InputStream is) {
+        if (md.isNeedWrap()) {
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Processor.java:49]: md 提供 unpack 是不是也合理，这块逻辑就能多协议复用了

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/ServerInboundObserver.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.common.extension.ExtensionLoader;
+import org.apache.dubbo.common.serialize.MultipleSerialization;
+import org.apache.dubbo.common.stream.StreamObserver;
+import org.apache.dubbo.config.Constants;
+import org.apache.dubbo.rpc.model.MethodDescriptor;
+import org.apache.dubbo.triple.TripleWrapper;
+
+public class ServerInboundObserver implements StreamObserver<Object> {
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/ServerInboundObserver.java:14]: 这个类没用到吗

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Stream.java [hunk 14]
```diff
@@ -25,7 +25,7 @@
 
     void onHeaders(Http2Headers headers);
 
-    void onData(InputStream in);
+    void onData(InputStream in) throws Exception;
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/Stream.java:28]: 应该在内部处理Exception

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamOutboundWriter.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
+package org.apache.dubbo.rpc.protocol.tri;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.dubbo.common.stream.StreamObserver;
+
+public class StreamOutboundWriter implements StreamObserver<Object> {
+
+    private StreamServerStream stream;
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamOutboundWriter.java:9]: final

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamServerStream.java [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    public void onError(GrpcStatus status) {
+    }
+
+    @Override
+    public void write(Object obj, ChannelPromise promise) throws Exception {
+        final Message message = (Message) obj;
+        final ByteBuf buf = getProcessor().encodeResponse(message, getCtx());
+        getCtx().write(new DefaultHttp2DataFrame(buf));
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamServerStream.java:85]: flush

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamServerStream.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    public void halfClose() throws Exception {
+        onComplete();
+    }
+
+
+    public void onComplete() {
+        final Http2Headers trailers = new DefaultHttp2Headers()
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamServerStream.java:94]: 需要判断 header /data/trailers 发送状态 避免异常时发送重复stream导致h2 error

**Review by guohao — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/stream/StreamObserver.java (hunk 3 above)
Comment (panxiaojun233) [reply dubbo-common/src/main/java/org/apache/dubbo/common/stream/StreamObserver.java:6]: fix

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java [hunk 18]
```diff
@@ -43,22 +47,19 @@
...
-    }
-
-    protected AbstractStream(URL url, ChannelHandlerContext ctx, boolean needWrap) {
+    protected AbstractStream(URL url, ChannelHandlerContext ctx, MethodDescriptor md) {
         this.ctx = ctx;
         this.url = url;
-        this.needWrap = needWrap;
-        if (needWrap) {
+        this.md = md;
+        this.datas = new MpscChunkedArrayQueue<>(16, 1 << 30);
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java:61]: 会有 MPSC 的情况吗，SPSC是不是就够了

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java [hunk 19]
```diff
@@ -110,12 +107,16 @@ public MultipleSerialization getMultipleSerialization() {
...
-            return;
+        //TODO requestN n>1 notify onNext(request)
+        try {
+            if (false) {
+                this.datas.add(in);
+            } else {
+                onSingleMessage(in);
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java:118]: Log or throw exception

File: dubbo-common/src/main/java/org/apache/dubbo/common/stream/StreamObserver.java [hunk 20]
```diff
@@ -0,0 +1,9 @@
+package org.apache.dubbo.common.stream;
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/stream/StreamObserver.java:1]: Please add Apache License Header

File: dubbo-common/src/main/java/org/apache/dubbo/rpc/model/MethodDescriptor.java [hunk 21]
```diff
@@ -90,4 +134,19 @@ public boolean isGeneric() {
         return generic;
     }
 
+    public enum RpcType {
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/rpc/model/MethodDescriptor.java:137]: What about extract it as a independent class alone with MethodDescriptor class?

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamServerStream.java [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+        getCtx().writeAndFlush(new DefaultHttp2DataFrame(buf));
+    }
+
+    public void halfClose() {
+        onComplete();
+    }
+
+
+    public void onComplete() {
+        // todo 需要判断 header /data/trailers 发送状态 避免异常时发送重复stream导致h2 error
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamServerStream.java:92]: pls comment in English

File: dubbo-common/src/main/java/org/apache/dubbo/rpc/model/MethodDescriptor.java [hunk 23]
```diff
@@ -40,21 +44,61 @@
...
+    }
+
+    public boolean isNeedWrap() {
+        return rpcType.equals(RpcType.UNARY_WRAP);
+    }
+
+    private boolean needWrap() {
+        if (CommonConstants.$INVOKE.equals(methodName) || CommonConstants.$INVOKE_ASYNC.equals(methodName)) {
+            return true;
+        } else if ("$echo".equals(methodName)) {
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/rpc/model/MethodDescriptor.java:87]: "$echo" should be replaced with constant

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java [hunk 24]
```diff
@@ -177,6 +187,20 @@ protected void convertAttachment(Http2Headers trailers, Map<String, Object> atta
                     trailers.add(key + "-tw-bin", encoded);
                 }
             }
+        } catch (IOException e) {
+            // todo log
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/AbstractStream.java:191]: FIX

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamOutboundWriter.java [hunk 25]
```diff
@@ -0,0 +1,27 @@
+package org.apache.dubbo.rpc.protocol.tri;
+
+import org.apache.dubbo.common.stream.StreamObserver;
+
+public class StreamOutboundWriter implements StreamObserver<Object> {
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamOutboundWriter.java:5]: <T>

Comment (guohao) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamOutboundWriter.java:5]: <T>

Comment (guohao) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/StreamOutboundWriter.java:5]:
```suggestion
public class StreamOutboundWriter implements StreamObserver<T> {
```

**Review by guohao — CHANGES_REQUESTED** [review]

**Review by guohao — APPROVED** [review]

**Pull request merged**